model.py

```diff
@@ -249,33 +267,54 @@ def decode_mae(self, x, ids_restore):
         x = x[:, 1:, :]
         return x
 
+    def encode_causal(self, x, mask_ratio):
+        x = self.conv1(x)
+        x = x.reshape((x.shape[0], -1, x.shape[-1]))
+        T = x.shape[1]
+        x = x + self.positional_embedding[1:, :].astype(x.dtype)
+
+        len_keep = int(x.shape[1] * (1 - mask_ratio))
+        x = x[:, :len_keep]
+
+        x = self.fwd_transformer(x)
+        return x, T
+
+    def decode_causal(self, x, T):
+        x = self.decoder_embed(x)
+        mask_tokens = jnp.tile(
+            self.mask_token,
+            (x.shape[0], T - x.shape[1] + 1, 1)
+        )
+        x = jnp.concatenate([x, mask_tokens], axis=1)
+
+        x = x + self.decoder_positional_embedding.astype(x.dtype)
+        x = self.dec_ln_pre(x)
+        attn_mask = nn.make_causal_mask(jnp.ones((x.shape[0], T + 1), dtype=x.dtype))
+        x = self.causal_decoder(x, attn_mask)
+        x = self.dec_ln_post(x)
+        x = self.decoder_pred(x)
+        return x[:, 1:, :]
+
     def perturb(self, x, noise_std, rng):
-        noise = jax.random.normal(rng, shape=x.shape, dtype=x.dtype)
-        x = x + noise_std * noise
+        noise = noise_std * jax.random.normal(rng, shape=x.shape, dtype=x.dtype)
```

Comment:
`noise` is separated from `noise_std` as is described in this paper (should confirm that though):

https://arxiv.org/abs/2205.11423

Reply:
Yeah but in the loss we were supervising MSE between the standard normal noise & the scaled noise, so wasn't sure. Shouldn't really matter since it's just a single constant scale factor that'll be learned

Reply:
Hmmm... makes sense but I think we should investigate. It's weird enough that they choose to regress the noise rather than the pixel values. There are some ablations in that paper that we should revisit.